File: lorenz-attractor.py
#!/usr/bin/env python3
import csv
from pathlib import Path
import mido
from mido import Message, MidiFile, MidiTrack

midi_range = (48, 83)
x_range = (-25.0, 25.0)
yz_range = (0.0, 50.0)

SIGMA = 10.0
RHO = 28.0
BETA = 8.0 / 3.0

def uniform_scaling(source, target, value):
    a, b = source
    c, d = target
    if b == a:
        return c
    return c + (value - a) * (d - c) / (b - a)

def rk4(f, t, y, dt):
    k1 = f(t, y)
    k2 = f(t + dt / 2, [y[i] + k1[i] * dt / 2 for i in range(len(y))])
    k3 = f(t + dt / 2, [y[i] + k2[i] * dt / 2 for i in range(len(y))])
    k4 = f(t + dt, [y[i] + k3[i] * dt for i in range(len(y))])
    return [
        y[i] + (dt / 6) * (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i])
        for i in range(len(y))
    ]

def lorenz(t, y):
    x, yy, z = y
    return [
        SIGMA * (yy - x),
        x * (RHO - z) - yy,
        x * yy - BETA * z,
    ]

def avoid_accidental(value):
    n = int(round(value))
    naturals = {0, 2, 4, 5, 7, 9, 11}
    while (n % 12) not in naturals:
        n += 1
    return n

def create_midi(notes, path):
    mid = MidiFile()
    track = MidiTrack()
    mid.tracks.append(track)
    track.append(mido.MetaMessage('track_name', name='Lorenz attractor', time=0))
    track.append(Message('program_change', program=4, time=0))
    time_per_chord = 480

    for chord in notes:
        for note in chord:
            track.append(Message('note_on', note=note, velocity=64, time=0))
        track.append(Message('note_off', note=chord[0], velocity=64, time=time_per_chord))
        track.append(Message('note_off', note=chord[1], velocity=64, time=0))
        track.append(Message('note_off', note=chord[2], velocity=64, time=0))

    mid.save(path)


def main():
    t = 0.0
    t_end = 50.0
    dt = 0.01
    y = [1.0, 1.0, 1.0]

    path = Path(__file__)
    csv_path = path.with_suffix(path.suffix + '.csv')
    mid_path = path.with_suffix(path.suffix + '.mid')

    notes = []
    with csv_path.open('w', newline='') as fh:
        writer = csv.writer(fh)
        writer.writerow(['t', 'x', 'y', 'z'])
        while t <= t_end:
            writer.writerow([f'{t:.8g}', f'{y[0]:.8g}', f'{y[1]:.8g}', f'{y[2]:.8g}'])
            y = rk4(lorenz, t, y, dt)
            t += dt
            n1 = uniform_scaling(x_range, midi_range, y[0])
            n2 = uniform_scaling(yz_range, midi_range, y[1])
            n3 = uniform_scaling(yz_range, midi_range, y[2])
            notes.append((avoid_accidental(n1), avoid_accidental(n2), avoid_accidental(n3)))

    create_midi(notes, mid_path)

if __name__ == '__main__':
    main()

File: lorenz-attractor.pl
#!/usr/bin/env perl
use v5.36;

use Math::Utils qw(uniform_scaling);
use MIDI::Util qw(setup_score);
use Music::Note ();

# ── Lorenz parameters ──────────────────────────────────────────────
use constant SIGMA => 10;
use constant RHO   => 28;
use constant BETA  => 8/3;

# ── MIDI ──────────────────────────-------------────────────────────
my $midi_range = [48, 83];  # allowed MIDI pitches
my $x_range    = [-25, 25]; # x-axis scale
my $yz_range   = [0, 50];   # y- & z-axis scale

my $score = setup_score(patch => 4);

# ── RK4 integrator ────────────────────────────────────────────────
sub vec_add_scale ($i, $j, $s) { [ map { $i->[$_] + $j->[$_] * $s } 0 .. $#$i ] }
sub rk4 ($f, $t, $y, $dt) {
    my $k1 = $f->($t, $y);
    my $k2 = $f->($t + $dt/2, vec_add_scale($y, $k1, $dt/2));
    my $k3 = $f->($t + $dt/2, vec_add_scale($y, $k2, $dt/2));
    my $k4 = $f->($t + $dt, vec_add_scale($y, $k3, $dt));
    return [
        map {
            $y->[$_] + ($dt/6) * ($k1->[$_] + 2*$k2->[$_] + 2*$k3->[$_] + $k4->[$_])
        } 0 .. $#$y
    ];
}

# ── Lorenz system ─────────────────────────────────────────────────
my $lorenz = sub ($t, $y) {
    my ($x, $yy, $z) = @$y;
    return [
        SIGMA * ($yy - $x),
        $x * (RHO - $z) - $yy,
        $x * $yy - BETA * $z,
    ]
};

# ── Initial conditions ────────────────────────────────────────────
my $t     = 0.0;
my $t_end = 50.0;
my $dt    = 0.01;
my $y     = [1.0, 1.0, 1.0];    # initial [x, y, z]

# ── Solve ----------------─────────────────────────────────────────
open my $fh, '>', "$0.csv" or die "Cannot open csv: $!";
say $fh 't,x,y,z';

while ($t <= $t_end) {
    say $fh join ',', map { sprintf '%.8g', $_ } $t, @$y;
    $y = rk4($lorenz, $t, $y, $dt);
    $t += $dt;
    # XXX naive
    my $n1 = sprintf '%.0f', uniform_scaling($x_range, $midi_range, $y->[0]);
    my $note = Music::Note->new($n1, 'midinum');
    $n1++ if $note->format('isobase') =~ /[#b]/;
    my $n2 = sprintf '%.0f', uniform_scaling($yz_range, $midi_range, $y->[1]);
    $note = Music::Note->new($n2, 'midinum');
    $n2++ if $note->format('isobase') =~ /[#b]/;
    my $n3 = sprintf '%.0f', uniform_scaling($yz_range, $midi_range, $y->[2]);
    $note = Music::Note->new($n3, 'midinum');
    $n3++ if $note->format('isobase') =~ /[#b]/;
    # say "N: $n1, $n2, $n3";
    $score->n('qn', $n1, $n2, $n3);
}

close $fh;
$score->write_score("$0.mid");
